resources/tag_models/event.json

```diff
@@ -0,0 +1,29 @@
+{
+  "tag": {
+	"api_key": "", // String - API key - Make sure the request is from a valid client (idea nicked from Keen.io) useful if a page gets copied onto a Russian website and creates noise
+	"version": "", // Version of the tracking client e.g. "JS v1.2"
+	"type": "event", // Type of request (page or event)
+	"id": "", // ID of this request, must be unique
+	"offset": 0 // Delay of this tag between event happening and being sent to server - milliseconds
+  },
+
+  "event": {
+	"page_id": "", // ID of the page on which this event happened.
+	"parent_id": "", // Parent event ID if this is a child event.
```

Comment:
@triblondon 

The collation ID's were a nice idea, but we've thought about it further and we feel we can't do it.

To be able to store that in a database it would be a many to many relationship. e.g.

Event table

id | type | data
---- | ---- | ----
1 | page | { ... }
2 | event | { ... }
3 | event | { ... } 

Collation table

event_id | collation_id
----------- | ---------------
1 | 2
1 | 3

The most common query would be: "what events happened on a page?". To answer that question it has to do 2 joins, whereas with the "parent, child" hierarchy it only needs to do one. 2 joins is bad from a performance point of view, but probably also from an analytics person who's reporting on the data PoV as it's more complex.

The performance issue is made worse when you consider we estimate to have 5.5 billion events per year (more with the help of origami). We also estimate an event will have on average, 2 collation ID's, sometimes more, sometimes less. That makes the collation table at 11 billion rows per year.

We will want want to keep 3 years of data, which will make the collation table 30-40 billion rows long.

So, we think it's not going to be very performant doing many-to-many joins over that table and therefore prefer the hierarchy model. It's not easy to turn the collation model into the hierarchy model later on as how do you tell which is the correct or most important collation ID?


Also, I don't follow the argument it is confusing to developers, these aren't "javascript" (or programatic) events and child events (or any kind of inheritance), it's just another piece of data to supply with an event.

To avoid the memory leak issue, I believe the best idea it to create a separate o-tracking guid component which gens guids for components to use when sending events.